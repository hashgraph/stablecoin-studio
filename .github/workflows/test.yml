name: test
on:
  pull_request:
    branches: [ main, release/**, feature/** ]
  push:
    branches: [ main, release/** ]
    tags: [ v* ]
jobs:
  test-node:
    name: Unit Tests
    runs-on: ubuntu-latest
    steps:
      - name: Use Node.js TLS 16.x
        uses: actions/setup-node@v1
        with:
          node-version: 16.x
      - name: Checkout repository
        uses: actions/checkout@v2
      - name: Install dependencies
        run: npm i
      - name: Build
        run: npm run build
      - name: Test
        run: npm run test