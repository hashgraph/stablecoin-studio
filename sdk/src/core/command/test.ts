/* eslint-disable @typescript-eslint/no-empty-interface */

interface CommandResponse {}
interface ICommand {}

export class Command<T> implements ICommand {
	private $resultType!: T;
}

export interface ICommandBus {
	execute<X>(command: Command<X>): Promise<X>;
}

export type IInferringCommandHandler<CommandType extends Command<unknown>> =
	CommandType extends Command<infer ResultType>
		? {
				execute(command: CommandType): Promise<ResultType>;
		  }
		: never;

class ConcreteCommandResponse implements CommandResponse {
	constructor(public readonly status: string) {}
}

class ConcreteCommand extends Command<ConcreteCommandResponse> {
	constructor(public readonly value: string) {
		super();
	}
}

class CommandHandler<CommandType extends Command<unknown>>
	implements IInferringCommandHandler<ConcreteCommand>
{
	execute(command: ConcreteCommand): Promise<ConcreteCommandResponse> {
		throw new Error('Method not implemented.');
	}
}

class CommandBus implements ICommandBus {
	execute<X>(command: Command<X>): Promise<X> {
		throw new Error('Method not implemented.');
	}
}

// Test

const bus = new CommandBus();
const res = bus.execute(new ConcreteCommand('1'));
