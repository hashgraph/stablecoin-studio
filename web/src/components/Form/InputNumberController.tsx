import type { ComponentType } from 'react';
import { useState } from 'react';
import type {
	InputProps as ChakraInputProps,
	FormControlProps,
	FormLabelProps,
} from '@chakra-ui/react';
import {
	Button,
	Flex,
	FormControl,
	FormErrorMessage,
	Input,
	InputGroup,
	InputRightElement,
	Stack,
} from '@chakra-ui/react';
import type { Control, FieldValues, UseControllerProps } from 'react-hook-form';
import { Controller } from 'react-hook-form';
import type { NumericFormatProps } from 'react-number-format';
import { NumericFormat } from 'react-number-format';
import Icon from '../Icon';
import InputLabel from './InputLabel';
import type { InputAttributes } from 'react-number-format/types/types';

/**
 * NumberFormatValues copied from 'react-number-format' since it isn't exported from the library right now
 */
interface NumberFormatValues {
	floatValue: number | undefined;
	formattedValue: string;
	value: string;
}
export interface InputNumberControllerProps
	extends Omit<
			ChakraInputProps,
			'name' | 'width' | 'value' | 'defaultValue' | 'color' | 'height' | 'size' | 'type'
		>,
		NumericFormatProps {
	rules?: UseControllerProps['rules'];
	control: Control<FieldValues>;
	label?: string;
	disabled?: boolean;
	onChangeAux?: (values: NumberFormatValues) => void;
	onBlurAux?: (floatValue: NumberFormatValues['floatValue']) => void;
	showErrors?: boolean;
	'data-testid'?: string;
	formStyle?: FormControlProps;
	labelStyle?: FormLabelProps;
	name: string;
	hasArrows?: boolean;
	maxValue?: number;
	minValue?: number;
}

const InputNumberController = ({
	control,
	name,
	id,
	label = '',
	labelStyle,
	rules,
	isRequired = false,
	placeholder,
	onChangeAux,
	onBlurAux,
	maxValue,
	minValue,
	allowNegative = false,
	suffix,
	disabled,
	decimalSeparator = '.',
	thousandSeparator = ',',
	hasArrows = false,
	showErrors = true,
	'data-testid': dataTestId = name,
	formStyle,
	decimalScale = 2,
	...props
}: InputNumberControllerProps) => {
	const [inputFloatValue, setInputFloatValue] = useState<number>();

	return (
		<Controller
			control={control}
			rules={rules}
			name={name}
			render={({ field: { onChange, value }, fieldState: { error } }) => {
				const onChangeCustom = (values: NumberFormatValues) => {
					onChange(values.floatValue === undefined ? '' : values.floatValue);
					setInputFloatValue(values?.floatValue!);
					onChangeAux && onChangeAux(values);
				};

				const onBlurCustom = () => {
					onBlurAux && onBlurAux(inputFloatValue);
				};

				const getIsAllowedValue = (values: NumberFormatValues) => {
					if (maxValue === 0) return false;

					if (maxValue && minValue) {
						const { floatValue = 0 } = values;

						return minValue <= floatValue && floatValue <= maxValue;
					}

					if (maxValue) {
						const { floatValue = 0 } = values;

						return floatValue <= maxValue;
					}

					if (minValue) {
						const { floatValue = 0 } = values;

						return floatValue >= minValue;
					}
					return true;
				};

				const incrementValue = () => {
					if (maxValue) {
						maxValue !== value &&
							onChangeCustom({
								value: !value ? '1' : (value++).toString(),
								floatValue: !value ? 1 : value++,
								formattedValue: value,
							});
					} else {
						onChangeCustom({
							value: !value ? '1' : (value++).toString(),
							floatValue: !value ? 1 : value++,
							formattedValue: value,
						});
					}
				};

				const decrementValue = () => {
					if (minValue || minValue === 0) {
						minValue !== value &&
							onChangeCustom({
								value: !value ? minValue.toString() : (value--).toString(),
								floatValue: !value ? minValue : value--,
								formattedValue: value,
							});
					} else {
						onChangeCustom({
							value: !value ? '0' : (value--).toString(),
							floatValue: !value ? 0 : value--,
							formattedValue: value,
						});
					}
				};

				return (
					<Stack w='full'>
						<FormControl data-testid='formControl' isInvalid={!!error} id={id} {...formStyle}>
							{label && (
								<InputLabel isRequired={isRequired} style={labelStyle}>
									{label}
								</InputLabel>
							)}
							<InputGroup>
								<NumericFormat
									data-testid={dataTestId}
									suffix={suffix}
									value={value}
									displayType='input'
									type='text'
									thousandSeparator={thousandSeparator}
									decimalSeparator={decimalSeparator}
									placeholder={placeholder}
									onValueChange={(values) => onChangeCustom(values)}
									onBlur={onBlurCustom}
									isAllowed={getIsAllowedValue}
									allowNegative={allowNegative}
									disabled={disabled}
									customInput={Input as ComponentType<InputAttributes>}
									decimalScale={decimalScale}
									{...props}
								/>

								<InputRightElement w='auto'>
									{hasArrows && (
										<Flex>
											<Button variant='unstyled' w='20px' onClick={incrementValue}>
												<Icon name='CaretUp' w={6} />
											</Button>
											<Button variant='unstyled' w='20px' onClick={decrementValue}>
												<Icon name='CaretDown' w={6} />
											</Button>
										</Flex>
									)}
								</InputRightElement>
							</InputGroup>
							{showErrors && error && (
								<FormErrorMessage data-testid='input-error-message'>
									{error.message}
								</FormErrorMessage>
							)}
						</FormControl>
					</Stack>
				);
			}}
		/>
	);
};

export default InputNumberController;
