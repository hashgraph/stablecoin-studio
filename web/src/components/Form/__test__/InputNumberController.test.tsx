import { useForm, UseFormReturn } from 'react-hook-form';
import { act, fireEvent, waitFor } from '@testing-library/react';
import InputNumberController, { InputNumberControllerProps } from '../InputNumberController';
import { render } from '../../../test/index';

let form: UseFormReturn;

const RenderWithForm = (props: Omit<InputNumberControllerProps, 'control'>) => {
	const localForm = useForm({
		mode: 'onChange',
		criteriaMode: 'all',
	});

	form = localForm;

	return <InputNumberController control={localForm.control} {...props} />;
};

const defaultProps: Omit<InputNumberControllerProps, 'control'> = {
	name: 'input-test',
	'data-testid': 'input',
};

const factoryComponent = (props: Partial<InputNumberControllerProps> = defaultProps) => {
	return render(<RenderWithForm name='input-test' {...props} />);
};

describe(`<${InputNumberController.name} />`, () => {
	describe('.render', () => {
		test('should render the component', () => {
			const component = factoryComponent({ label: 'test-label' });

			expect(component.asFragment()).toMatchSnapshot('WithLabel');
		});

		test('should show the label', () => {
			const component = factoryComponent({ label: 'test-label' });

			expect(component.getByText('test-label')).toBeInTheDocument();
		});

		test('should show the error message', async () => {
			const component = factoryComponent({ showErrors: true });

			const errorMessage = 'Expired';
			act(() => {
				form.setError(defaultProps.name, { type: 'expired', message: errorMessage });
			});

			expect(component.getByTestId('input-error-message')).toHaveTextContent(errorMessage);
			expect(component.asFragment()).toMatchSnapshot('WithError');
		});
		test('shouldnt show the error message if showErrors=false', async () => {
			const component = factoryComponent({ showErrors: false });

			const errorMessage = 'Expired';
			act(() => {
				form.setError(defaultProps.name, { type: 'expired', message: errorMessage });
			});

			expect(component.queryByTestId('input-error-message')).toBeNull();
		});
		test('should show placeholder', () => {
			const component = factoryComponent({
				placeholder: 'placeholder',
			});

			const input = component.getByTestId(defaultProps.name);
			expect(input.getAttribute('placeholder')).toBe('placeholder');
			expect(component.asFragment()).toMatchSnapshot('WithPlaceholder');
		});

		test('should be disabled if disabled=true', () => {
			const component = factoryComponent({
				disabled: true,
			});
			const input = component.getByTestId(defaultProps.name);

			expect(input).toBeDisabled();
			expect(component.asFragment()).toMatchSnapshot('Disabled');
		});
	});
	describe('.events', () => {
		test('calls onBlurAux', async () => {
			const onBlurCustom = jest.fn();
			const component = factoryComponent({ onBlurAux: onBlurCustom });
			const input = component.getByTestId(defaultProps.name);

			fireEvent.blur(input);
			await waitFor(() => {
				expect(onBlurCustom).toHaveBeenCalledTimes(1);
			});
		});

		test('calls onChangeAux', async () => {
			const onChangeCustom = jest.fn();

			const component = factoryComponent({
				onChangeAux: onChangeCustom,
			});
			const input = component.getByTestId(defaultProps.name);

			fireEvent.change(input, { target: { value: '123' } });

			await waitFor(() => {
				expect(onChangeCustom).toHaveBeenCalled();
			});
		});
	});
});
