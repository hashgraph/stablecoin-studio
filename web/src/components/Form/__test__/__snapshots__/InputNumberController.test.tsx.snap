// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InputNumberController /> .render should be disabled if disabled=true: Disabled 1`] = `
<DocumentFragment>
  <div
    class="chakra-stack css-1h0sumw"
  >
    <div
      class="chakra-form-control css-1kxonj9"
      data-testid="formControl"
      role="group"
    >
      <div
        class="chakra-input__group css-4302v8"
      >
        <input
          class="chakra-input css-16lr1fk"
          data-testid="input-test"
          disabled=""
          id="field-:r5:"
          inputmode="numeric"
          type="text"
          value=""
        />
        <div
          class="chakra-input__right-element css-y1u7d2"
        />
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<InputNumberController /> .render should render the component: WithLabel 1`] = `
<DocumentFragment>
  <div
    class="chakra-stack css-1h0sumw"
  >
    <div
      class="chakra-form-control css-1kxonj9"
      data-testid="formControl"
      role="group"
    >
      <label
        class="chakra-form__label css-1y854pg"
        for="field-:r0:"
        id="field-:r0:-label"
      >
        <div
          class="chakra-stack css-84zodg"
        >
          <p
            class="chakra-text css-mgupmj"
          >
            test-label
          </p>
        </div>
      </label>
      <div
        class="chakra-input__group css-4302v8"
      >
        <input
          class="chakra-input css-16lr1fk"
          data-testid="input-test"
          id="field-:r0:"
          inputmode="numeric"
          type="text"
          value=""
        />
        <div
          class="chakra-input__right-element css-y1u7d2"
        />
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<InputNumberController /> .render should show placeholder: WithPlaceholder 1`] = `
<DocumentFragment>
  <div
    class="chakra-stack css-1h0sumw"
  >
    <div
      class="chakra-form-control css-1kxonj9"
      data-testid="formControl"
      role="group"
    >
      <div
        class="chakra-input__group css-4302v8"
      >
        <input
          class="chakra-input css-16lr1fk"
          data-testid="input-test"
          id="field-:r4:"
          inputmode="numeric"
          placeholder="placeholder"
          type="text"
          value=""
        />
        <div
          class="chakra-input__right-element css-y1u7d2"
        />
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<InputNumberController /> .render should show the error message: WithError 1`] = `
<DocumentFragment>
  <div
    class="chakra-stack css-1h0sumw"
  >
    <div
      class="chakra-form-control css-1kxonj9"
      data-testid="formControl"
      role="group"
    >
      <div
        class="chakra-input__group css-4302v8"
      >
        <input
          aria-describedby="field-:r2:-feedback"
          aria-invalid="true"
          class="chakra-input css-16lr1fk"
          data-testid="input-test"
          id="field-:r2:"
          inputmode="numeric"
          type="text"
          value=""
        />
        <div
          class="chakra-input__right-element css-y1u7d2"
        />
      </div>
      <div
        aria-live="polite"
        class="chakra-form__error-message css-2e3jkk"
        data-testid="input-error-message"
        id="field-:r2:-feedback"
      >
        Expired
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
